apps/studio/src/server/modules/page/page.router.ts

```diff
@@ -100,6 +110,84 @@ export const pageRouter = router({
         ...siteMeta,
       }
     }),
+
+  reorderBlock: protectedProcedure
```

Comment:
suspect this whole function should be wrapped in a transaction. We want to ensure that there is no read/write skew when we are validating the content.

Reply:
could you elaborate in abit more detail? not too sure on the portion re: read/write skew - my interpretation is that someone updates between the read and our write.

if that's the case, i'm actually ok not wrapping in tx because i suspect that the time between read/write is going to be so minute (no async calls, pure CPU + it's mostly just sanity checks) that this occurs v rarely

Reply:
skew (which i realize just wrapping transactions don't guard against) can occur when:

1. Person 1 reorders block from A to A1 state
  - this function fetch full page A
2. Person 2 reorders block from A to A2 state
  - fetch full page A
  - equality check succeeds
  - page A -> A2 state
3. P1's reorder somehow slower than P2
  - equality check still succeeds, since P1 fetched state A before P2's reordering finished
  - page A2 -> A1 state (invalid, but allowed)

Should have been thrown instead of this invalid state. May cause drifting yeah.

Suggest using a transaction with a [`FOR UPDATE`](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS) clause.

this can be done in kysely just by adding a `.forUpdate` in the query clause when fetching the full page.

Reply:
i added the `forUpdate` clause! not very sure i understand how it works - read your link and [this](https://www.cockroachlabs.com/blog/select-for-update/) here, it seems like it's basically locking a row for read?

in our case above, wouldn't the read still succeed then the write fail cos they're separate? lets clarify this tomorrow maybe i grug brain